File: 783_minimum_distance_between_bst_nodes/minimum_distance_between_bst_nodes.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def iter(self, root: TreeNode, result: list):
        if not root:
            return
        result.append(root.val)
        self.iter(root.left, result)
        self.iter(root.right, result)

    def minDiffInBST(self, root: TreeNode) -> int:
        values = []
        self.iter(root, values)
        values.sort()
        ans = values[1] - values[0]
        for i in range(2, len(values)):
            ans = ans if values[i] - values[i - 1] > ans else values[i] - values[i - 1]
        return ans


if __name__ == '__main__':
    tree = TreeNode(4, TreeNode(2, TreeNode(1), TreeNode(3)), TreeNode(6))
    print(Solution().minDiffInBST(tree))

File: 783_minimum_distance_between_bst_nodes/minimum_distance_between_bst_nodes.java
import java.util.ArrayList;
import java.util.Collections;

class Solution {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    private void prevIter(TreeNode node, ArrayList<Integer> result) {
        if (node == null) {
            return;
        }
        result.add(node.val);
        prevIter(node.left, result);
        prevIter(node.right, result);
    }

    public int minDiffInBST(TreeNode root) {
        ArrayList<Integer> arr = new ArrayList<Integer>();
        prevIter(root, arr);
        Collections.sort(arr);
        int ans = arr.get(1) - arr.get(0);
        for (int i = 2; i < arr.size(); i ++) {
            ans = arr.get(i) - arr.get(i-1) < ans ? arr.get(i) - arr.get(i-1) : ans;
        }
        return ans;
    }
}